ros/src/computing/planning/motion/packages/astar_planner/nodes/velocity_set/velocity_set.cpp

```diff
@@ -384,95 +376,102 @@ int detectDecelerateObstacle(const pcl::PointCloud<pcl::PointXYZ>& points, const
 {
-  int stop_obstacle_waypoint = -1;
-  EObstacleType obstacle_type = EObstacleType::NONE;
-
   // no input for detection || no closest waypoint
   if ((points.empty() == true && vs_info.getDetectionResultByOtherNodes() == -1) || closest_waypoint < 0)
   {
     // equal NONE
   }
   else
   {
-    stop_obstacle_waypoint =
-      detectStopObstacle(points, closest_waypoint, lane, crosswalk, vs_info.getStopRange(),
-                         vs_info.getPointsThreshold(), vs_info.getLocalizerPose(),
-                         obstacle_points, &obstacle_type, vs_info.getDetectionResultByOtherNodes());
+    detectStopObstacle(points, closest_waypoint, lane, crosswalk, vs_info.getStopRange(),
+                       vs_info.getPointsThreshold(), vs_info.getLocalizerPose(),
+                       obstacle_infos, vs_info.getDetectionResultByOtherNodes());
   }
 
-  // tracking vehicle on waypoints
-  if (obstacle_type == EObstacleType::ON_WAYPOINTS)
+  // tracking "nearest" vehicle ON_WAYPOINTS
+  bool obstacle_is_observed = false;
+  for (auto& info : *obstacle_infos)
   {
-    double waypoint_velocity = lane.waypoints.at(stop_obstacle_waypoint).twist.twist.linear.x;
-    tracker->update(stop_obstacle_waypoint, obstacle_points, waypoint_velocity, vs_info.getControlPose().pose.position);
-    *obstacle_waypoint = tracker->getWaypointIdx();
-    *obstacle_velocity = tracker->getVelocity();
-  }
-  else if (obstacle_type == EObstacleType::NONE)
-  {
-    tracker->update();
-    *obstacle_waypoint = tracker->getWaypointIdx();
-    *obstacle_velocity = tracker->getVelocity();
+    if (info.type == EObstacleType::ON_WAYPOINTS)
+    {
+      double waypoint_velocity = lane.waypoints.at(info.waypoint).twist.twist.linear.x;
+      tracker->update(info.waypoint, &info.points, waypoint_velocity, vs_info.getControlPose().pose.position);
+      info.waypoint = tracker->getWaypointIdx();
+      info.velocity = tracker->getVelocity();
+      obstacle_is_observed = true;
+      break;
+    }
   }
-  else
+
+  // obstacle ON_WAYPOINTS is not observed
+  if (obstacle_infos->size() == 0 || !obstacle_is_observed)
   {
     tracker->update();
-    *obstacle_waypoint = stop_obstacle_waypoint;
-    *obstacle_velocity = 0.0;
+    if (tracker->getWaypointIdx() != -1)  // tracking is continued
+    {
+      obstacle_infos->push_back(ObstacleInfo(tracker->getWaypointIdx(),
+        tracker->getVelocity(), EObstacleType::ON_WAYPOINTS));
+    }
   }
 
   // skip searching deceleration range
   if (vs_info.getDecelerationRange() < 0.01)
   {
-    if (*obstacle_waypoint < 0)
+    // return EControl based on nearest object
+    if (obstacle_infos->size() == 0)
       return EControl::KEEP;
-    else if (obstacle_type == EObstacleType::ON_WAYPOINTS || obstacle_type == EObstacleType::ON_CROSSWALK || obstacle_type == EObstacleType::NONE)
```

Comment:
why did you remove EObstacleType::NONE?

Reply:
you have no EObstacleType::NONE obstacle, it's just for initialization.